compiler/rustc_resolve/src/late.rs

```diff
@@ -181,7 +189,8 @@ pub(crate) enum RibKind<'a> {
     FnOrCoroutine,
 
     /// We passed through an item scope. Disallow upvars.
-    Item(HasGenericParams),
+    /// Note if this is a static/const to help with some confusing diagnostics
+    Item(HasGenericParams, IsStaticOrConst),
```

Comment:
I wonder if we can just record the item's `DefKind` in the rib. Seems a bit more general than just recording if it's a static or const. We should have access to that at this point, probably?